src/net/dns.cc

```diff
@@ -806,81 +806,106 @@ class net::dns_resolver::impl
-                    f = e.tcp.out->write(std::move(p));
-                    break;
-                case type::udp:
-                    f = e.udp.channel.send(e.udp.dst, std::move(p));
-                    break;
-                default:
-                    return -1;
-                }
+            return send_packet(fd, std::move(p));
+        } catch (...) {
+        }
+        return -1;
+    }
 
-                if (!f.available()) {
-                    dns_log.trace("Send {} unavailable.", fd);
-                    e.avail &= ~POLLOUT;
-                    use(fd);
-                    // FIXME: future is discarded
-                    (void)f.then_wrapped([me = shared_from_this(), &e, bytes, fd](future<> f) {
-                        try {
-                            f.get();
-                            dns_log.trace("Send {}. {} bytes sent.", fd, bytes);
-                        } catch (...) {
-                            dns_log.debug("Send {} failed: {}", fd, std::current_exception());
-                        }
-                        e.avail |= POLLOUT;
-                        me->poll_sockets();
-                        me->release(fd);
-                    });
-                    // c-ares does _not_ use non-blocking retry for udp sockets. We just pretend
-                    // all is fine even though we have no idea. Barring stack/adapter failure it
-                    // is close to the same guarantee a "normal" message send would have anyway.
-                    // For tcp we also pretend we're done, to make sure we don't have to deal with
-                    // matching sent data
-                    return bytes;
-                }
-                if (f.failed()) {
-                    try {
-                        f.get();
-                    } catch (std::system_error& e) {
-                        errno = e.code().value();
-                    } catch (...) {
-                    }
-                    return -1;
-                }
+    ssize_t send_packet(ares_socket_t fd, net::packet p) {
+        auto& e = _sockets.at(fd);
+        auto bytes = p.len();
+        auto f = make_ready_future();
 
+        switch (e.typ) {
+        case type::tcp:
+            if (!e.tcp.out) {
+                e.tcp.out = e.tcp.socket.output(0);
+            }
+            f = e.tcp.out->write(std::move(p));
+            break;
+        case type::udp:
+            // #1285 - serialize UDP sends if we are overlapping with a pending send.
+            if (!(e.avail & POLLOUT)) {
+                auto op = std::exchange(_send_waiters[fd], promise<>()); 
+                (void)_send_waiters[fd].get_future().then([me = shared_from_this(), op = std::move(op), fd, p = std::move(p), this]() mutable {
+                    op.set_value(); // make sure we signal chain.
+                    dns_log.trace("Send {} retry...", fd);
+                    send_packet(fd, std::move(p)); // try again.
+                });
```

Comment:
This is overkill, if _send_waiters[fd] is a future<> you can write

```c++
auto& chain = _send_waiters.at[fd];
chain = chain.then_wrapped(.... [] { return e.tcp.out->write(std::move(p)); });
```

btw I see no error handling on failed writes.

Reply:
If there's an error we can mark the socket as broken and then return an error on the next send.

Though can a udp socket really get poisoned this way? Probably not (e.g. an error in a udp socket doesn't affect the next send, unlike tcp).

Reply:
If the socket would be so broken, it will fast-fail me thinks. All else should/would be packet errors down the line...